File: compatibility-tests/sdk-test-java/src/test/java/com/floci/test/KmsFeaturesTest.java
package com.floci.test;

import org.junit.jupiter.api.*;
import software.amazon.awssdk.services.kms.KmsClient;
import software.amazon.awssdk.services.kms.model.CreateKeyResponse;
import software.amazon.awssdk.services.kms.model.GetKeyPolicyResponse;
import software.amazon.awssdk.services.kms.model.ListResourceTagsResponse;

import java.util.Map;

import static org.assertj.core.api.Assertions.*;

/**
 * Compatibility tests for KMS fixes:
 *   #269 — CreateKey applies Tags at creation time
 *   #258 — GetKeyPolicy returns the stored policy
 *   #259 — PutKeyPolicy updates the key policy
 */
@DisplayName("KMS features (#258 #259 #269)")
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class KmsFeaturesTest {

    private static KmsClient kms;

    @BeforeAll
    static void setup() {
        kms = TestFixtures.kmsClient();
    }

    @AfterAll
    static void cleanup() {
        if (kms != null) kms.close();
    }

    // ── Issue #269 — CreateKey applies Tags ───────────────────────────────────

    @Test
    @Order(10)
    void createKeyWithTagsStoresTags() {
        CreateKeyResponse resp = kms.createKey(b -> b
                .description("tagged-key")
                .tags(
                        software.amazon.awssdk.services.kms.model.Tag.builder().tagKey("env").tagValue("prod").build(),
                        software.amazon.awssdk.services.kms.model.Tag.builder().tagKey("team").tagValue("platform").build()
                ));
        String keyId = resp.keyMetadata().keyId();

        ListResourceTagsResponse tags = kms.listResourceTags(b -> b.keyId(keyId));
        Map<String, String> tagMap = tags.tags().stream()
                .collect(java.util.stream.Collectors.toMap(
                        software.amazon.awssdk.services.kms.model.Tag::tagKey,
                        software.amazon.awssdk.services.kms.model.Tag::tagValue));

        assertThat(tagMap).containsEntry("env", "prod");
        assertThat(tagMap).containsEntry("team", "platform");

        kms.scheduleKeyDeletion(b -> b.keyId(keyId).pendingWindowInDays(7));
    }

    @Test
    @Order(11)
    void createKeyWithoutTagsHasEmptyTagList() {
        CreateKeyResponse resp = kms.createKey(b -> b.description("no-tags-key"));
        String keyId = resp.keyMetadata().keyId();

        ListResourceTagsResponse tags = kms.listResourceTags(b -> b.keyId(keyId));
        assertThat(tags.tags()).isEmpty();

        kms.scheduleKeyDeletion(b -> b.keyId(keyId).pendingWindowInDays(7));
    }

    // ── Issue #258 — GetKeyPolicy ─────────────────────────────────────────────

    @Test
    @Order(20)
    void createKeyWithoutPolicyReturnsDefaultPolicy() {
        CreateKeyResponse resp = kms.createKey(b -> b.description("default-policy-key"));
        String keyId = resp.keyMetadata().keyId();

        GetKeyPolicyResponse policyResp = kms.getKeyPolicy(b -> b.keyId(keyId));
        assertThat(policyResp.policy()).isNotBlank();
        assertThat(policyResp.policyName()).isEqualTo("default");
        assertThat(policyResp.policy()).contains("kms:*");

        kms.scheduleKeyDeletion(b -> b.keyId(keyId).pendingWindowInDays(7));
    }

    @Test
    @Order(21)
    void createKeyWithPolicyStoresAndReturnsPolicy() {
        String customPolicy = "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"Custom\"," +
                "\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::000000000000:root\"}," +
                "\"Action\":\"kms:*\",\"Resource\":\"*\"}]}";

        CreateKeyResponse resp = kms.createKey(b -> b
                .description("custom-policy-key")
                .policy(customPolicy));
        String keyId = resp.keyMetadata().keyId();

        GetKeyPolicyResponse policyResp = kms.getKeyPolicy(b -> b.keyId(keyId));
        assertThat(policyResp.policy()).isEqualTo(customPolicy);
        assertThat(policyResp.policyName()).isEqualTo("default");

        kms.scheduleKeyDeletion(b -> b.keyId(keyId).pendingWindowInDays(7));
    }

    // ── Issue #259 — PutKeyPolicy ─────────────────────────────────────────────

    @Test
    @Order(30)
    void putKeyPolicyUpdatesPolicy() {
        CreateKeyResponse resp = kms.createKey(b -> b.description("put-policy-key"));
        String keyId = resp.keyMetadata().keyId();

        String newPolicy = "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"Updated\"," +
                "\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::000000000000:root\"}," +
                "\"Action\":\"kms:Decrypt\",\"Resource\":\"*\"}]}";

        kms.putKeyPolicy(b -> b.keyId(keyId).policy(newPolicy));

        GetKeyPolicyResponse policyResp = kms.getKeyPolicy(b -> b.keyId(keyId));
        assertThat(policyResp.policy()).isEqualTo(newPolicy);

        kms.scheduleKeyDeletion(b -> b.keyId(keyId).pendingWindowInDays(7));
    }

    @Test
    @Order(31)
    void putKeyPolicyRoundTrip() {
        CreateKeyResponse resp = kms.createKey(b -> b.description("round-trip-key"));
        String keyId = resp.keyMetadata().keyId();

        // Get initial policy
        String initial = kms.getKeyPolicy(b -> b.keyId(keyId)).policy();
        assertThat(initial).isNotBlank();

        // Put a new policy
        String updated = "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"RoundTrip\"," +
                "\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::000000000000:root\"}," +
                "\"Action\":\"kms:*\",\"Resource\":\"*\"}]}";
        kms.putKeyPolicy(b -> b.keyId(keyId).policy(updated));

        // Verify change persisted
        assertThat(kms.getKeyPolicy(b -> b.keyId(keyId)).policy()).isEqualTo(updated);
        assertThat(kms.getKeyPolicy(b -> b.keyId(keyId)).policy()).isNotEqualTo(initial);

        kms.scheduleKeyDeletion(b -> b.keyId(keyId).pendingWindowInDays(7));
    }
}

File: compatibility-tests/sdk-test-node/tests/kms-features.test.ts
/**
 * Compatibility tests for KMS fixes:
 *   #269 — CreateKey applies Tags at creation time
 *   #258 — GetKeyPolicy returns the stored policy
 *   #259 — PutKeyPolicy updates the key policy
 */

import { describe, it, expect, afterEach } from 'vitest';
import {
  KMSClient,
  CreateKeyCommand,
  GetKeyPolicyCommand,
  PutKeyPolicyCommand,
  ListResourceTagsCommand,
  ScheduleKeyDeletionCommand,
} from '@aws-sdk/client-kms';
import { makeClient } from './setup';

const kms = makeClient(KMSClient);
const createdKeyIds: string[] = [];

async function createKey(opts: Parameters<typeof CreateKeyCommand>[0] = {}) {
  const resp = await kms.send(new CreateKeyCommand(opts));
  const keyId = resp.KeyMetadata!.KeyId!;
  createdKeyIds.push(keyId);
  return resp;
}

afterEach(async () => {
  // Schedule deletion for all keys created in the test
  while (createdKeyIds.length > 0) {
    const keyId = createdKeyIds.pop()!;
    try {
      await kms.send(new ScheduleKeyDeletionCommand({ KeyId: keyId, PendingWindowInDays: 7 }));
    } catch { /* ignore */ }
  }
});

// ── Issue #269 — CreateKey applies Tags ───────────────────────────────────────

describe('KMS features (#258 #259 #269)', () => {

  it('#269: CreateKey with Tags stores tags immediately', async () => {
    const resp = await createKey({
      Description: 'tagged-key',
      Tags: [
        { TagKey: 'env', TagValue: 'prod' },
        { TagKey: 'team', TagValue: 'platform' },
      ],
    });
    const keyId = resp.KeyMetadata!.KeyId!;

    const tags = await kms.send(new ListResourceTagsCommand({ KeyId: keyId }));
    const tagMap = Object.fromEntries(tags.Tags!.map(t => [t.TagKey!, t.TagValue!]));

    expect(tagMap['env']).toBe('prod');
    expect(tagMap['team']).toBe('platform');
  });

  it('#269: CreateKey without Tags has empty tag list', async () => {
    const resp = await createKey({ Description: 'no-tags-key' });
    const keyId = resp.KeyMetadata!.KeyId!;

    const tags = await kms.send(new ListResourceTagsCommand({ KeyId: keyId }));
    expect(tags.Tags).toHaveLength(0);
  });

  // ── Issue #258 — GetKeyPolicy ───────────────────────────────────────────────

  it('#258: CreateKey without Policy returns a non-empty default policy', async () => {
    const resp = await createKey({ Description: 'default-policy-key' });
    const keyId = resp.KeyMetadata!.KeyId!;

    const policyResp = await kms.send(new GetKeyPolicyCommand({ KeyId: keyId }));
    expect(policyResp.Policy).toBeTruthy();
    expect(policyResp.PolicyName).toBe('default');
    expect(policyResp.Policy).toContain('kms:*');
  });

  it('#258: CreateKey with Policy stores and returns that policy', async () => {
    const customPolicy = JSON.stringify({
      Version: '2012-10-17',
      Statement: [{
        Sid: 'Custom',
        Effect: 'Allow',
        Principal: { AWS: 'arn:aws:iam::000000000000:root' },
        Action: 'kms:*',
        Resource: '*',
      }],
    });

    const resp = await createKey({ Description: 'custom-policy-key', Policy: customPolicy });
    const keyId = resp.KeyMetadata!.KeyId!;

    const policyResp = await kms.send(new GetKeyPolicyCommand({ KeyId: keyId }));
    expect(policyResp.Policy).toBe(customPolicy);
    expect(policyResp.PolicyName).toBe('default');
  });

  // ── Issue #259 — PutKeyPolicy ───────────────────────────────────────────────

  it('#259: PutKeyPolicy updates the key policy', async () => {
    const resp = await createKey({ Description: 'put-policy-key' });
    const keyId = resp.KeyMetadata!.KeyId!;

    const newPolicy = JSON.stringify({
      Version: '2012-10-17',
      Statement: [{
        Sid: 'Updated',
        Effect: 'Allow',
        Principal: { AWS: 'arn:aws:iam::000000000000:root' },
        Action: 'kms:Decrypt',
        Resource: '*',
      }],
    });

    await kms.send(new PutKeyPolicyCommand({ KeyId: keyId, Policy: newPolicy }));

    const policyResp = await kms.send(new GetKeyPolicyCommand({ KeyId: keyId }));
    expect(policyResp.Policy).toBe(newPolicy);
  });

  it('#259: PutKeyPolicy round-trip — get, change, verify', async () => {
    const resp = await createKey({ Description: 'round-trip-key' });
    const keyId = resp.KeyMetadata!.KeyId!;

    const initial = (await kms.send(new GetKeyPolicyCommand({ KeyId: keyId }))).Policy!;
    expect(initial).toBeTruthy();

    const updated = JSON.stringify({
      Version: '2012-10-17',
      Statement: [{
        Sid: 'RoundTrip',
        Effect: 'Allow',
        Principal: { AWS: 'arn:aws:iam::000000000000:root' },
        Action: 'kms:*',
        Resource: '*',
      }],
    });

    await kms.send(new PutKeyPolicyCommand({ KeyId: keyId, Policy: updated }));

    const after = (await kms.send(new GetKeyPolicyCommand({ KeyId: keyId }))).Policy!;
    expect(after).toBe(updated);
    expect(after).not.toBe(initial);
  });
});
